predicators/approaches/initialized_pg3_approach.py

```diff
@@ -98,20 +107,58 @@ class _Analogy:
     # All maps are base -> target.
     predicates: Dict[Predicate, Predicate]
     nsrts: Dict[NSRT, NSRT]
-    types: Dict[Type, Type]
+    # Keep track of variables on a per-NSRT basis because some variable names
+    # might be the same across multiple NSRTs, but we don't want to match them.
+    nsrt_variables: Dict[Tuple[NSRT, Variable], Tuple[NSRT, Variable]]
+
+    @cached_property
+    def types(self) -> Dict[Type, Type]:
+        """Infer type analogy from variables."""
+        type_map: Dict[Type, Type] = {}
+        for nsrt in self.nsrts:
+            var_map = self.base_nsrt_to_variable_analogy(nsrt)
+            for old_var, new_var in var_map.items():
+                type_map[old_var.type] = new_var.type
```

Comment:
Should we make sure that the SME result consistently matches the types, e.g., base variables of type A and matched exclusively to target variables of type B? I'm not sure if that would happen often, but SME is generally not type-aware. [I can make it, though, that wouldn't be too difficult]

Reply:
I think that suffices

Reply:
I started to go down this road, but it's a bit annoying to test because we don't have domains that actually have different types yet (since ferry and gripper just have 'object'). and in fact, we might want to think about just converting all types to unary predicates.... so for now I'll leave it alone